build: update jasmine to 3.5

JiaLiPassion opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


Update jasmine to 3.5, so I can continue to work with #33657, also need to wait #33717 typescript 3.7 PR to merge first, otherwise, it will not compile.

JiaLiPassion committed:
build: update jasmine to 3.5

1. update jasmine to 3.5
2. update @types/jasmine to 3.5
3. update @types/jasminewd2 to 2.0.8

Also fix several cases, the new jasmine 3 will help to create test cases correctly,
such as in the `jasmine 2.x` version, the following case will pass

```
expect(1 == 2);
```

But in jsamine 3, the case will need to be

```
expect(1 == 2).toBeTrue();
```

mhevery reviewed (commented):
I see this is still WIP

JiaLiPassion commented:
@mhevery , yeah, this is still WIP, I will finish this one asap.

JiaLiPassion commented:
@mhevery, I updated the PR, some case may need to wait for the `@types/jasmine` update, but most cases are ok for now, and not only the `jasmine` and `@types/jasmine` are updated, some previous wrong cases are also be found, such as https://github.com/angular/angular/pull/34625/files#diff-8783fc8fad0324187d972850b203c93cR84-R89, this case will not pass, I am not sure how to fix this one, could you take a look? Thanks.

gkalpak committed:
fixup! build: update jasmine to 3.5

gkalpak reviewed (changes requested):
I have pushed some small fixes/improvements in a fixup commit. Please take a look and let me know what you think, @JiaLiPassion.

Other than that lgtm for `elements`, `ngcc`, `service-worker`, `upgrade` (except we should have issues for `@types/jasmine` problems and reference them in the TODO comments).

Reviewed-for: fw-elements, fw-ngcc, fw-server-worker, fw-upgrade

gkalpak commented:
@googlebot I concent.

JiaLiPassion commented:
@gkalpak, thanks for the review.
The issues for `@types/jasmine` is here.
1. the issue of overloading error
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42455, it seems to be fixed by this one https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43004
2. the issue of optional method error
 https://github.com/DefinitelyTyped/DefinitelyTyped/issues/43486

gkalpak commented:
It seems that they decided against merging DefinitelyTyped/DefinitelyTyped#43004, so it doesn't seem we can expect a fix soon. 

Can please mention DefinitelyTyped/DefinitelyTyped#42455 and DefinitelyTyped/DefinitelyTyped#43486 in the TODO comments (so future readers know what issues you refer to)?

gkalpak commented:
Thx for the info, @JiaLiPassion!

It seems that they decided against merging DefinitelyTyped/DefinitelyTyped#43004, so it doesn't seem we can expect a fix soon. 

Can please mention DefinitelyTyped/DefinitelyTyped#42455 and DefinitelyTyped/DefinitelyTyped#43486 in the TODO comments (so future readers know what issues you refer to)?

JiaLiPassion commented:
@gkalpak, got it, sure, will add the issues to the comments.

JiaLiPassion committed:
fix(zone.js): add issue numbers of `@types/jasmine` to the test cases

Some cases will still need to use `spy as any` cast, because `@types/jasmine` have some issues,
1. The issue jasmine doesn't handle optional method properties, https://github.com/DefinitelyTyped/DefinitelyTyped/issues/43486
2. The issue jasmine doesn't handle overload method correctly, https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42455

gkalpak commented:
@JiaLiPassion, I see you removed my fixup commit (80970fe9d898f279360b131970b1840768b936ae). Was that intentional?
Also, I don't see the comments updated with issue numbers 😕

JiaLiPassion commented:
@gkalpak, sorry I wrongly pushed this PR on my another PC (without your commits and my fixes to add the issues numbers), please review the new commits, I will check my commits more carefully the next time, thank you.

gkalpak commented:
Cool, thx! I just wasn't sure whether you intentionally discarded my commit or whether it was accidental :grin:
I'll take a other look.

gkalpak commented on packages/zone.js/test/extra/bluebird.spec.ts:
```diff
@@ -284,11 +284,7 @@ describe('bluebird promise', () => {
                 expect(r).toBe(arr[idx]);
-=======
-                expect(r === arr[idx]).toBeTrue();
->>>>>>> 253023848d... build: update jasmine to 3.5
```
Merge conflict markers 😱 :grin:

JiaLiPassion replied:
Sorry, this has been resolved. Now it should ok.

gkalpak commented on packages/zone.js/test/extra/bluebird.spec.ts:
```diff
@@ -309,11 +305,7 @@ describe('bluebird promise', () => {
                 expect(r).toBe(arr[idx]);
-=======
-                expect(r === arr[idx]).toBeTrue();
->>>>>>> 253023848d... build: update jasmine to 3.5
```
Merge conflict markers 😱 :grin:

JiaLiPassion replied:
Sorry, this has been resolved. Now it should ok.

gkalpak commented on packages/core/test/render3/node_selector_matcher_spec.ts:
```diff
@@ -81,10 +81,9 @@ describe('css selector matching', () => {
+      it('should match namespaced attributes', () => {
         expect(isMatching(
-                   'span', [AttributeMarker.NamespaceURI, 'http://some/uri', 'title', 'name'],
+                   'span', ['title', 'name', AttributeMarker.NamespaceURI, 'http://some/uri'],
```
This looks wrong. This is basically testing non-namespaced attributes. It might be that the test was wrong and it should assert that namespaced attributes are _NOT_ matched, but someone familiar with this part of the code needs to confirm.

In any case, swapping the order fo the attributes is not the right way to del with it :grin:

JiaLiPassion replied:
Yeah, I am not sure about this one, @mhevery, could you check this one? thanks.

mhevery replied:
OK, I think we need to revert that, and figure out why the test fails. My guess is that this test never worked.  I tried to look into the implementation but was not able to figure out the issue without spending a lot of time on it. I am fine with `xit` and a comment since it never worked.

JiaLiPassion replied:
Got it, I will revert it and mark it as `xit` for now.

gkalpak commented on packages/core/test/render3/node_selector_matcher_spec.ts:
```diff
@@ -81,10 +81,13 @@ describe('css selector matching', () => {
 
-      it('should match namespaced attributes', () => {
+      // TODO: this case will not work, need more discussion
+      // https://github.com/angular/angular/pull/34625/files#r401853803
```
This link does not work for me? Maybe try:

```suggestion
      // https://github.com/angular/angular/pull/34625#discussion_r401791275
```

JiaLiPassion replied:
Got it, thanks, updated.

gkalpak reviewed (approved):
LGTM (with one minor comment) 👍
Great work, @JiaLiPassion ✨

JiaLiPassion commented:
@gkalpak , thanks for the review again, I see you add the `cleanup` label, is that for the `namespace match` case? I just updated it to `xit`.

gkalpak commented:
It's about https://github.com/angular/angular/pull/34625#discussion_r401862922.

JiaLiPassion commented:
@gkalpak , got it, thanks, just updated.

JiaLiPassion commented:
@gkalpak , I pushed another commit to resolve the `clang format` issues, it seems in this PR, when I run `yarn gulp format`, a lot of files' format are changed such as this one, https://github.com/angular/angular/pull/34625/files/e9033e63103c0306a69a854acad1018bf6b79575..0ba68c4a4645b6f87068cd800b8c6e84cea05e23#diff-5cf554492fad6b4ee0121404631404a9R32, but in `master` branch, the same `clang-format` version `1.4.0`, no files need to be changed, not sure why. Could you help to check the new commit is valid ? Thank you.

gkalpak reviewed (approved):
Context on why these files are formatted:

Clang-format has been updated to a newer version, which formats files differently. In order to avoid a huge commit that would touch most files in the repo (and mess with git history/blaming), the formatting logic has changed so that formatting is only enforced on files that are changed in a PR.

So, all files touched in this PR also need to be formatted by the new clang-format version. This is expected.

JiaLiPassion commented:
@gkalpak , got it, thank you for the explanation.

atscott reviewed (approved):
Reviewed-for: fw-router

AndrewKushnir commented:
Hi,

Based on PullApprove, I was added as a reviewer on behalf of `fw-core`. Misko (also a member of `fw-core`) already approved this PR, so my LGTM is not needed, so I removed myself from the list of reviewers.

Note: I've added "cleanup" label because there is a couple files that conflict with master branch.

Thank you.

atscott reviewed (approved):
Reviewed-for: fw-router

IgorMinar reviewed (approved):
thanks  @JiaLiPassion !

I'm relabeling this as master & patch because this change doesn't affect Angular applications.

@JiaLiPassion I doubt that this change will cherry-pick cleanly to 9.1.x branch - can you please send a separate PR against that branch? thanks!

IgorMinar commented:
@JiaLiPassion please rebase this PR one more time - I just spoke with @kara and we'll merge this PR as soon as it's rebased.

JiaLiPassion commented:
@IgorMinar , thanks for the review, I just rebased, and I will create another PR for 9.1.x branch.
